### PR TITLE
Option to disable printing package-name in public or private interface.

### PR DESCRIPTION
**Pull request by elsh**
Having `-package-name` flag in non-package interfaces causes package-only dependencies to be built even for an external client, and causes a missing dependency module error. 

For example, when building X that depends on A with the following dependency chain, where A, B, and C are in the same package, and X is an external client (client outside of that package):
  X --> A --> B --(package-only)--> C

1. When dependency scanning X, and opening up A and B, because the scan target is in a different package domain, the scanner decides that B's package-only dependency on C is to be ignored.
2. When then finally building A itself, it will load its dependencies, but because the .private.swiftinterface of A still specifies `-package-name`, when it loads B, it will then examine its dependencies and deem that this package-only dependency on C is required.

Because (1) and (2) disagree, we get a `error: missing required module 'C'` now when building the private A textual interface.

This PR provides an option to disable printing package-name flag in public or private swiftinterface, so in the case of (2) above, it no longer requires package-only dependency module. 

rdar://130701866

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

File: test/ModuleInterface/package_interface_disable_package_name.swift [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+// RUN:   -enable-library-evolution -swift-version 6 \
+// RUN:   -package-name barpkg \
+// RUN:   -Xfrontend -disable-print-package-name-for-non-package-interface \
+// RUN:   -emit-module-interface-path %t/Bar.swiftinterface \
+// RUN:   -emit-private-module-interface-path %t/Bar.private.swiftinterface \
+// RUN:   -emit-package-module-interface-path %t/Bar.package.swiftinterface
+
+// RUN: %FileCheck %s --check-prefix=CHECK-PUBLIC < %t/Bar.swiftinterface
+// RUN: %FileCheck %s --check-prefix=CHECK-PRIVATE < %t/Bar.private.swiftinterface
+// RUN: %FileCheck %s --check-prefix=CHECK-PACKAGE < %t/Bar.package.swiftinterface
```
Comment (nkcsgexi) [inline test/ModuleInterface/package_interface_disable_package_name.swift:15]: Could you also test that after dropping `-package-name foopkg`, we can still compile the interface without errors?

File: test/ModuleInterface/package_interface_disable_package_name.swift [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+// CHECK-PUBLIC-NOT: -package-name foopkg
+// CHECK-PUBLIC-NOT: -package-name barpkg
+// CHECK-PRIVATE-NOT: -package-name foopkg
+// CHECK-PRIVATE-NOT: -package-name barpkg
+// CHECK-PACKAGE-NOT: -package-name foopkg
+// CHECK-PACKAGE: -package-name barpkg
```
Comment (nkcsgexi) [inline test/ModuleInterface/package_interface_disable_package_name.swift:22]: We could also test that in all cases, flags like `-module-name Bar` and `-swift-version 6` won't be dropped.

**Comment by aschwaighofer** [comment]
Why do we remove the string after the fact rather than preventing it from being added in the first place.

It feels like the logic in `SaveModuleInterfaceArgs` can be made to only add the string to `.package.swiftinterface` files?

File: test/ModuleInterface/package_interface_disable_package_name.swift (hunk 1 above)
Comment (elsh) [reply test/ModuleInterface/package_interface_disable_package_name.swift:15]: done

**Comment by elsh** [comment]
> Why do we remove the string after the fact rather than preventing it from being added in the first place.
> 
> It feels like the logic in `SaveModuleInterfaceArgs` can be made to only add the string to `.package.swiftinterface` files?

When `SaveModuleInterfaceArgs` is called, the interface mode is unknown. Later, `printModuleInterfaceIfNeeded` is called per mode (public, private, or package), and that's when the disabling logic can be applied. See `emitSwiftInterface` which is called by `printModuleInterfaceIfNeeded`  and also has other per-mode specific logic. 

```
  if (opts.InputsAndOutputs.hasModuleInterfaceOutputPath()) {
    hadAnyError |= printModuleInterfaceIfNeeded(...);
  }

  if (opts.InputsAndOutputs.hasPrivateModuleInterfaceOutputPath()) {
    ModuleInterfaceOptions privOpts = Invocation.getModuleInterfaceOptions();
    privOpts.setInterfaceMode(PrintOptions::InterfaceMode::Private);
    privOpts.ModulesToSkipInPublicInterface.clear();
    hadAnyError |= printModuleInterfaceIfNeeded(...);
  }
  if (opts.InputsAndOutputs.hasPackageModuleInterfaceOutputPath()) {
    ModuleInterfaceOptions pkgOpts = Invocation.getModuleInterfaceOptions();
    pkgOpts.setInterfaceMode(PrintOptions::InterfaceMode::Package);
    pkgOpts.ModulesToSkipInPublicInterface.clear();
    hadAnyError |= printModuleInterfaceIfNeeded(...);
  }

```

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by aschwaighofer** [comment]
I don't think you have to refactor the code. You can follow the existing pattern of `SaveModuleInterfaceArgs`.

You would add a `std::string FlagsForPackageOnly` to `ModuleInterfaceOptions`.

https://github.com/swiftlang/swift/blob/3ed1cf379131cbb9c6712b9e787472c606cb904a/include/swift/Frontend/ModuleInterfaceSupport.h#L48


In `SaveModuleInterfaceArgs` you would add the option to either `Flag` or `FlagsForPackageOnly` depending on whether `DisablePackageNameForNonPackageInterface` is false or true.

In `printToolVersionAndFlagsComment` you add the content of `FlagsForPackageOnly` only to the package interface.

**Comment by elsh** [comment]
> FlagsForPackageOnly

Thanks for the suggestion! Will update accordingly.

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test

**Comment by elsh** [comment]
@swift-ci smoke test macOS

**Comment by elsh** [comment]
@swift-ci smoke test macOS

**Pull request merged**